README.md
# Doodles

## Setup My PC
- [setup_my_pc](https://github.com/JaehyunShim/doodles/blob/master/setup_my_pc.md)

## Study Programming
- [study_cpp](https://github.com/JaehyunShim/doodles/blob/master/study_cpp.md)
- [study_python](https://github.com/JaehyunShim/doodles/blob/master/study_python.md)
- [study_cmake](https://github.com/JaehyunShim/doodles/blob/master/study_cmake.md)
- [study_qt](https://github.com/JaehyunShim/doodles/blob/master/study_qt.md)
- [study_markdown](https://gist.github.com/ihoneymon/652be052a0727ad59601)
- [study_euslisp](https://github.com/JaehyunShim/doodles/blob/master/study_euslisp.md)
- [study_js](https://github.com/JaehyunShim/doodles/blob/master/study_js.md)

## Study Robotics
- [study_ros](https://github.com/JaehyunShim/doodles/blob/master/study_ros.md)
- [study_control](https://github.com/JaehyunShim/doodles/blob/master/study_control.md)
- [study_wheeled](https://github.com/JaehyunShim/doodles/blob/master/study_wheeled.md)
- [study_manipulator](https://github.com/JaehyunShim/doodles/blob/master/study_manipulator.md)

## Study DevOps
- [study_git](https://github.com/JaehyunShim/doodles/blob/master/study_git.md)
- [study_coding_style](https://github.com/JaehyunShim/doodles/blob/master/study_coding_style.md)
- [study_ci](https://github.com/JaehyunShim/doodles/blob/master/study_ci.md)
- [study_docker](https://github.com/JaehyunShim/doodles/blob/master/study_docker.md)
- [study_unit_test](https://github.com/JaehyunShim/doodles/blob/master/study_unit_test.md)
- [study_documentation](https://github.com/JaehyunShim/doodles/blob/master/study_documentation.md)
- [study_bloom](https://github.com/JaehyunShim/doodles/blob/master/study_bloom.md)

## Study Others
- [study_intro2cs](https://github.com/JaehyunShim/doodles/blob/master/study_intro2cs.md)
- [study_linux](https://github.com/JaehyunShim/doodles/blob/master/study_linux.md)
- [study_rtos](https://github.com/JaehyunShim/doodles/blob/master/study_rtos.md)
- [study_remote_pc_control](https://github.com/JaehyunShim/doodles/blob/master/study_remote_pc_control.md)
- [study_usb](https://github.com/JaehyunShim/doodles/blob/master/study_usb.md)
- [study_nvidia](https://github.com/JaehyunShim/doodles/blob/master/study_nvidia.md)
- [study_jekyll](https://github.com/JaehyunShim/doodles/blob/master/study_jekyll.md)


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package doodles
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.0 (2020-01-01)
------------------
* added packages for doodles
* Contributors: Jaehyun Shim


install_ros2_foxy.sh
#!/bin/bash
# Apache License 2.0
# Copyright (c) 2020, ROBOTIS CO., LTD.

echo ""
echo "[Note] OS version  >>> Ubuntu 20.04 (Focal Fossa) or Linux Mint 21.x"
echo "[Note] Target ROS version >>> ROS 2 Foxy Fitzroy"
echo "[Note] Colcon workspace   >>> $HOME/colcon_ws"
echo ""
echo "PRESS [ENTER] TO CONTINUE THE INSTALLATION"
echo "IF YOU WANT TO CANCEL, PRESS [CTRL] + [C]"
read

echo "[Set the target ROS version and name of colcon workspace]"
name_ros_version=${name_ros_version:="foxy"}
name_colcon_workspace=${name_colcon_workspace:="colcon_ws"}

echo "[Setup Locale]"
sudo locale-gen en_US en_US.UTF-8
sudo update-locale LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8
export LANG=en_US.UTF-8

echo "[Setup Sources]"
sudo rm -rf /var/lib/apt/lists/* && sudo apt update && sudo apt install -y curl gnupg2 lsb-release
curl -s https://raw.githubusercontent.com/ros/rosdistro/master/ros.asc | sudo apt-key add -
sudo sh -c 'echo "deb http://packages.ros.org/ros2/ubuntu `lsb_release -cs` main" > /etc/apt/sources.list.d/ros2-latest.list'

echo "[Install ROS 2 packages]"
sudo apt update && sudo apt install -y ros-$name_ros_version-desktop

echo "[Environment setup]"
source /opt/ros/$name_ros_version/setup.sh
sudo apt install -y python3-argcomplete python3-colcon-common-extensions python3-vcstool

echo "[Make the colcon workspace and test colcon build]"
mkdir -p $HOME/$name_colcon_workspace/src
cd $HOME/$name_colcon_workspace
colcon build --symlink-install

echo "[Set the ROS evironment]"
sh -c "echo \"alias gb='gedit ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias nb='nano ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias sb='source ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias gs='git status'\" >> ~/.bashrc"
sh -c "echo \"alias gp='find . -maxdepth 3 -name .git -type d | rev | cut -c 6- | rev | xargs -I {} git -C {} pull'\" >> ~/.bashrc"
sh -c "echo \"alias nt='nautilus .'\" >> ~/.bashrc"

sh -c "echo \"alias cw='cd ~/$name_colcon_workspace'\" >> ~/.bashrc"
sh -c "echo \"alias cs='cd ~/$name_colcon_workspace/src'\" >> ~/.bashrc"
sh -c "echo \"alias cb='cd ~/$name_colcon_workspace && colcon build --symlink-install && source ~/$name_colcon_workspace/install/local_setup.bash'\" >> ~/.bashrc"
sh -c "echo \"alias ct='cd ~/$name_colcon_workspace && colcon test'\" >> ~/.bashrc"
sh -c "echo \"alias ctr='cd ~/$name_colcon_workspace && colcon test-result --verbose'\" >> ~/.bashrc"
sh -c "echo \"alias ci='cd ~/$name_catkin_workspace && rosdep update && rosdep install --from-paths src --ignore-src --rosdistro $name_ros_version -r -y'\" >> ~/.bashrc"
sh -c "echo \"alias cr='cd ~/$name_colcon_workspace && rm -rf build install log'\" >> ~/.bashrc"

sh -c "echo \"alias pe='printenv | grep ROS'\" >> ~/.bashrc"

sh -c "echo \"alias ac='ament_cpplint'\" >> ~/.bashrc"
sh -c "echo \"alias af='ament_flake8'\" >> ~/.bashrc"

sh -c "echo \"alias ka='killall -9'\" >> ~/.bashrc"
sh -c "echo \"alias kg='killall -9 gazebo & killall -9 gzserver  & killall -9 gzclient'\" >> ~/.bashrc"

sh -c "echo \"source /opt/ros/$name_ros_version/setup.bash\" >> ~/.bashrc"
sh -c "echo \"source ~/$name_colcon_workspace/install/local_setup.bash\" >> ~/.bashrc"

sh -c "echo \"export ROS_DOMAIN_ID=30\" >> ~/.bashrc"
sh -c "echo \"export TURTLEBOT3_MODEL=burger\" >> ~/.bashrc"
sh -c "echo \"alias tb3='ssh ubuntu@0.0.0.0' \" >> ~/.bashrc"

exec bash

echo "[Complete!!!]"
exit 0


install_ros_kinetic.sh
#!/bin/bash
# Apache License 2.0
# Copyright (c) 2017, ROBOTIS CO., LTD.

echo ""
echo "[Note] Target OS version  >>> Ubuntu 16.04.x (xenial) or Linux Mint 18.x"
echo "[Note] Target ROS version >>> ROS Kinetic Kame"
echo "[Note] Catkin workspace   >>> $HOME/catkin_ws"
echo ""
echo "PRESS [ENTER] TO CONTINUE THE INSTALLATION"
echo "IF YOU WANT TO CANCEL, PRESS [CTRL] + [C]"
read

echo "[Set the target OS, ROS version and name of catkin workspace]"
name_os_version=${name_os_version:="xenial"}
name_ros_version=${name_ros_version:="kinetic"}
name_catkin_workspace=${name_catkin_workspace:="catkin_ws"}

echo "[Update the package lists and upgrade them]"
sudo apt-get update -y
sudo apt-get upgrade -y

echo "[Install build environment, the chrony, ntpdate and set the ntpdate]"
sudo apt-get install -y chrony ntpdate build-essential
sudo ntpdate ntp.ubuntu.com

echo "[Add the ROS repository]"
if [ ! -e /etc/apt/sources.list.d/ros-latest.list ]; then
  sudo sh -c "echo \"deb http://packages.ros.org/ros/ubuntu ${name_os_version} main\" > /etc/apt/sources.list.d/ros-latest.list"
fi

echo "[Download the ROS keys]"
roskey=`apt-key list | grep "Open Robotics"`
if [ -z "$roskey" ]; then
  sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-key C1CF6E31E6BADE8868B172B4F42ED6FBAB17C654
fi

echo "[Check the ROS keys]"
roskey=`apt-key list | grep "Open Robotics"`
if [ -n "$roskey" ]; then
  echo "[ROS key exists in the list]"
else
  echo "[Failed to receive the ROS key, aborts the installation]"
  exit 0
fi

echo "[Update the package lists and upgrade them]"
sudo apt-get update -y
sudo apt-get upgrade -y

echo "[Install the ros-desktop-full and all rqt plugins]"
sudo apt-get install -y ros-$name_ros_version-desktop-full ros-$name_ros_version-rqt-*

echo "[Initialize rosdep]"
sudo sh -c "rosdep init"
rosdep update

echo "[Environment setup and getting rosinstall]"
source /opt/ros/$name_ros_version/setup.sh
sudo apt-get install -y python-rosinstall

echo "[Make the catkin workspace and test the catkin_make]"
mkdir -p $HOME/$name_catkin_workspace/src
cd $HOME/$name_catkin_workspace/src
catkin_init_workspace
cd $HOME/$name_catkin_workspace
catkin_make

echo "[Set the ROS evironment]"
sh -c "echo \"alias gb='gedit ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias nb='nano ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias sb='source ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias gs='git status'\" >> ~/.bashrc"
sh -c "echo \"alias gp='find . -maxdepth 3 -name .git -type d | rev | cut -c 6- | rev | xargs -I {} git -C {} pull'\" >> ~/.bashrc"
sh -c "echo \"alias nt='nautilus .'\" >> ~/.bashrc"

sh -c "echo \"alias cw='cd ~/$name_catkin_workspace'\" >> ~/.bashrc"
sh -c "echo \"alias cs='cd ~/$name_catkin_workspace/src'\" >> ~/.bashrc"
sh -c "echo \"alias cm='cd ~/$name_catkin_workspace && catkin_make && rospack profile && source ~/$name_catkin_workspace/devel/setup.bash'\" >> ~/.bashrc"
sh -c "echo \"alias ct='cd ~/$name_catkin_workspace && catkin_make run_tests roslint && rospack profile && source ~/$name_catkin_workspace/devel/setup.bash'\" >> ~/.bashrc"
sh -c "echo \"alias ci='cd ~/$name_catkin_workspace && rosdep update && rosdep install --from-paths src --ignore-src --rosdistro $name_ros_version -r -y'\" >> ~/.bashrc"
sh -c "echo \"alias cr='cd ~/$name_catkin_workspace && rm -rf build devel'\" >> ~/.bashrc"

sh -c "echo \"alias pe='env | grep ROS'\" >> ~/.bashrc"

sh -c "echo \"alias ka='killall -9'\" >> ~/.bashrc"
sh -c "echo \"alias kg='killall -9 gazebo & killall -9 gzserver  & killall -9 gzclient'\" >> ~/.bashrc"

sh -c "echo \"source /opt/ros/$name_ros_version/setup.bash\" >> ~/.bashrc"
sh -c "echo \"source ~/$name_catkin_workspace/devel/setup.bash\" >> ~/.bashrc"

sh -c "echo \"export ROS_MASTER_URI=http://localhost:11311\" >> ~/.bashrc"
sh -c "echo \"export ROS_HOSTNAME=localhost\" >> ~/.bashrc"
sh -c "echo \"export TURTLEBOT3_MODEL=burger\" >> ~/.bashrc"
sh -c "echo \"alias tb3='ssh ubuntu@0.0.0.0' \" >> ~/.bashrc"

exec bash

echo "[Complete!!!]"
exit 0


install_ros_melodic.sh
#!/bin/bash
# Apache License 2.0
# Copyright (c) 2018, ROBOTIS CO., LTD.

echo ""
echo "[Note] Target OS version  >>> Ubuntu 18.04.x (Bionic Beaver) or Linux Mint 19.x"
echo "[Note] Target ROS version >>> ROS Melodic Morenia"
echo "[Note] Catkin workspace   >>> $HOME/catkin_ws"
echo ""
echo "PRESS [ENTER] TO CONTINUE THE INSTALLATION"
echo "IF YOU WANT TO CANCEL, PRESS [CTRL] + [C]"
read

echo "[Set the target OS, ROS version and name of catkin workspace]"
name_os_version=${name_os_version:="bionic"}
name_ros_version=${name_ros_version:="melodic"}
name_catkin_workspace=${name_catkin_workspace:="catkin_ws"}

echo "[Update the package lists and upgrade them]"
sudo apt update -y
sudo apt upgrade -y

echo "[Install build environment, the chrony, ntpdate and set the ntpdate]"
sudo apt install -y chrony ntpdate build-essential
sudo ntpdate ntp.ubuntu.com

echo "[Add the ROS repository]"
if [ ! -e /etc/apt/sources.list.d/ros-latest.list ]; then
  sudo sh -c "echo \"deb http://packages.ros.org/ros/ubuntu ${name_os_version} main\" > /etc/apt/sources.list.d/ros-latest.list"
fi

echo "[Download the ROS keys]"
roskey=`apt-key list | grep "Open Robotics"`
if [ -z "$roskey" ]; then
  sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-key C1CF6E31E6BADE8868B172B4F42ED6FBAB17C654
fi

echo "[Check the ROS keys]"
roskey=`apt-key list | grep "Open Robotics"`
if [ -n "$roskey" ]; then
  echo "[ROS key exists in the list]"
else
  echo "[Failed to receive the ROS key, aborts the installation]"
  exit 0
fi

echo "[Update the package lists and upgrade them]"
sudo apt update -y
sudo apt upgrade -y

echo "[Install the ros-desktop-full and all rqt plugins]"
sudo apt install -y ros-$name_ros_version-desktop-full ros-$name_ros_version-rqt-*

echo "[Initialize rosdep]"
sudo sh -c "rosdep init"
rosdep update

echo "[Environment setup and getting rosinstall]"
source /opt/ros/$name_ros_version/setup.sh
sudo apt install -y python-rosinstall python-rosinstall-generator python-wstool

echo "[Make the catkin workspace and test the catkin_make]"
mkdir -p $HOME/$name_catkin_workspace/src
cd $HOME/$name_catkin_workspace/src
catkin_init_workspace
cd $HOME/$name_catkin_workspace
catkin_make

echo "[Set the ROS evironment]"
sh -c "echo \"alias gb='gedit ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias nb='nano ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias sb='source ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias gs='git status'\" >> ~/.bashrc"
sh -c "echo \"alias gp='find . -maxdepth 3 -name .git -type d | rev | cut -c 6- | rev | xargs -I {} git -C {} pull'\" >> ~/.bashrc"
sh -c "echo \"alias nt='nautilus .'\" >> ~/.bashrc"

sh -c "echo \"alias cw='cd ~/$name_catkin_workspace'\" >> ~/.bashrc"
sh -c "echo \"alias cs='cd ~/$name_catkin_workspace/src'\" >> ~/.bashrc"
sh -c "echo \"alias cm='cd ~/$name_catkin_workspace && catkin_make && rospack profile && source ~/$name_catkin_workspace/devel/setup.bash'\" >> ~/.bashrc"
sh -c "echo \"alias ct='cd ~/$name_catkin_workspace && catkin_make run_tests roslint && rospack profile && source ~/$name_catkin_workspace/devel/setup.bash'\" >> ~/.bashrc"
sh -c "echo \"alias ci='cd ~/$name_catkin_workspace && rosdep update && rosdep install --from-paths src --ignore-src --rosdistro $name_ros_version -r -y'\" >> ~/.bashrc"
sh -c "echo \"alias cr='cd ~/$name_catkin_workspace && rm -rf build devel'\" >> ~/.bashrc"

sh -c "echo \"alias pe='env | grep ROS'\" >> ~/.bashrc"

sh -c "echo \"alias ka='killall -9'\" >> ~/.bashrc"
sh -c "echo \"alias kg='killall -9 gazebo & killall -9 gzserver  & killall -9 gzclient'\" >> ~/.bashrc"

sh -c "echo \"source /opt/ros/$name_ros_version/setup.bash\" >> ~/.bashrc"
sh -c "echo \"source ~/$name_catkin_workspace/devel/setup.bash\" >> ~/.bashrc"

sh -c "echo \"export ROS_MASTER_URI=http://localhost:11311\" >> ~/.bashrc"
sh -c "echo \"export ROS_HOSTNAME=localhost\" >> ~/.bashrc"
sh -c "echo \"export TURTLEBOT3_MODEL=burger\" >> ~/.bashrc"
sh -c "echo \"alias tb3='ssh ubuntu@0.0.0.0' \" >> ~/.bashrc"

exec bash

echo "[Complete!!!]"
exit 0


install_ros2_dashing.sh
#!/bin/bash
# Apache License 2.0
# Copyright (c) 2019, ROBOTIS CO., LTD.

echo ""
echo "[Note] OS version  >>> Ubuntu 18.04 (Bionic Beaver) or Linux Mint 19.x"
echo "[Note] Target ROS version >>> ROS 2 Dashing Diademata"
echo "[Note] Colcon workspace   >>> $HOME/colcon_ws"
echo ""
echo "PRESS [ENTER] TO CONTINUE THE INSTALLATION"
echo "IF YOU WANT TO CANCEL, PRESS [CTRL] + [C]"
read

echo "[Set the target ROS version and name of colcon workspace]"
name_ros_version=${name_ros_version:="dashing"}
name_colcon_workspace=${name_colcon_workspace:="colcon_ws"}

echo "[Setup Locale]"
sudo locale-gen en_US en_US.UTF-8
sudo update-locale LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8
export LANG=en_US.UTF-8

echo "[Setup Sources]"
sudo rm -rf /var/lib/apt/lists/* && sudo apt update && sudo apt install -y curl gnupg2 lsb-release
curl -s https://raw.githubusercontent.com/ros/rosdistro/master/ros.asc | sudo apt-key add -
sudo sh -c 'echo "deb [arch=amd64,arm64] http://packages.ros.org/ros2/ubuntu `lsb_release -cs` main" > /etc/apt/sources.list.d/ros2-latest.list'

echo "[Install ROS 2 packages]"
sudo apt update && sudo apt install -y ros-$name_ros_version-desktop

echo "[Environment setup]"
source /opt/ros/$name_ros_version/setup.sh
sudo apt install -y python3-argcomplete python3-colcon-common-extensions python3-vcstool

echo "[Make the colcon workspace and test colcon build]"
mkdir -p $HOME/$name_colcon_workspace/src
cd $HOME/$name_colcon_workspace
colcon build --symlink-install

echo "[Set the ROS evironment]"
sh -c "echo \"alias gb='gedit ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias nb='nano ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias sb='source ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias gs='git status'\" >> ~/.bashrc"
sh -c "echo \"alias gp='find . -maxdepth 3 -name .git -type d | rev | cut -c 6- | rev | xargs -I {} git -C {} pull'\" >> ~/.bashrc"
sh -c "echo \"alias nt='nautilus .'\" >> ~/.bashrc"

sh -c "echo \"alias cw='cd ~/$name_colcon_workspace'\" >> ~/.bashrc"
sh -c "echo \"alias cs='cd ~/$name_colcon_workspace/src'\" >> ~/.bashrc"
sh -c "echo \"alias cb='cd ~/$name_colcon_workspace && colcon build --symlink-install && source ~/$name_colcon_workspace/install/local_setup.bash'\" >> ~/.bashrc"
sh -c "echo \"alias ct='cd ~/$name_colcon_workspace && colcon test'\" >> ~/.bashrc"
sh -c "echo \"alias ctr='cd ~/$name_colcon_workspace && colcon test-result --verbose'\" >> ~/.bashrc"
sh -c "echo \"alias ci='cd ~/$name_catkin_workspace && rosdep update && rosdep install --from-paths src --ignore-src --rosdistro $name_ros_version -r -y'\" >> ~/.bashrc"
sh -c "echo \"alias cr='cd ~/$name_colcon_workspace && rm -rf build install log'\" >> ~/.bashrc"

sh -c "echo \"alias pe='printenv | grep ROS'\" >> ~/.bashrc"

sh -c "echo \"alias ac='ament_cpplint'\" >> ~/.bashrc"
sh -c "echo \"alias af='ament_flake8'\" >> ~/.bashrc"

sh -c "echo \"alias ka='killall -9'\" >> ~/.bashrc"
sh -c "echo \"alias kg='killall -9 gazebo & killall -9 gzserver  & killall -9 gzclient'\" >> ~/.bashrc"

sh -c "echo \"source /opt/ros/$name_ros_version/setup.bash\" >> ~/.bashrc"
sh -c "echo \"source ~/$name_colcon_workspace/install/local_setup.bash\" >> ~/.bashrc"

sh -c "echo \"export ROS_DOMAIN_ID=30\" >> ~/.bashrc"
sh -c "echo \"export TURTLEBOT3_MODEL=burger\" >> ~/.bashrc"
sh -c "echo \"alias tb3='ssh ubuntu@0.0.0.0' \" >> ~/.bashrc"

exec bash

echo "[Complete!!!]"
exit 0


install_ros_noetic.sh
#!/bin/bash
# Apache License 2.0
# Copyright (c) 2020, ROBOTIS CO., LTD.

echo ""
echo "[Note] Target OS version  >>> Ubuntu 20.04.x (Focal Fossa) or Linux Mint 21.x"
echo "[Note] Target ROS version >>> ROS Noetic Ninjemys"
echo "[Note] Catkin workspace   >>> $HOME/catkin_ws"
echo ""
echo "PRESS [ENTER] TO CONTINUE THE INSTALLATION"
echo "IF YOU WANT TO CANCEL, PRESS [CTRL] + [C]"
read

echo "[Set the target OS, ROS version and name of catkin workspace]"
name_os_version=${name_os_version:="focal"}
name_ros_version=${name_ros_version:="noetic"}
name_catkin_workspace=${name_catkin_workspace:="catkin_ws"}

echo "[Update the package lists and upgrade them]"
sudo apt update -y
sudo apt upgrade -y

echo "[Install build environment, the chrony, ntpdate and set the ntpdate]"
sudo apt install -y chrony ntpdate build-essential
sudo ntpdate ntp.ubuntu.com

echo "[Add the ROS repository]"
if [ ! -e /etc/apt/sources.list.d/ros-latest.list ]; then
  sudo sh -c "echo \"deb http://packages.ros.org/ros/ubuntu ${name_os_version} main\" > /etc/apt/sources.list.d/ros-latest.list"
fi

echo "[Download the ROS keys]"
roskey=`apt-key list | grep "Open Robotics"`
if [ -z "$roskey" ]; then
  sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-key C1CF6E31E6BADE8868B172B4F42ED6FBAB17C654
fi

echo "[Check the ROS keys]"
roskey=`apt-key list | grep "Open Robotics"`
if [ -n "$roskey" ]; then
  echo "[ROS key exists in the list]"
else
  echo "[Failed to receive the ROS key, aborts the installation]"
  exit 0
fi

echo "[Update the package lists and upgrade them]"
sudo apt update -y
sudo apt upgrade -y

echo "[Install the ros-desktop-full and all rqt plugins]"
sudo apt install -y ros-$name_ros_version-desktop-full
sudo apt-get install -y ros-$name_ros_version-rqt-*

echo "[Initialize rosdep]"
sudo sh -c "rosdep init"
rosdep update

echo "[Environment setup and getting rosinstall]"
source /opt/ros/$name_ros_version/setup.sh
sudo apt install -y python3-rosinstall python3-rosinstall-generator python3-wstool git

echo "[Make the catkin workspace and test the catkin_make]"
mkdir -p $HOME/$name_catkin_workspace/src
cd $HOME/$name_catkin_workspace/src
catkin_init_workspace
cd $HOME/$name_catkin_workspace
catkin_make

echo "[Set the ROS evironment]"
sh -c "echo \"alias gb='gedit ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias nb='nano ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias sb='source ~/.bashrc'\" >> ~/.bashrc"
sh -c "echo \"alias gs='git status'\" >> ~/.bashrc"
sh -c "echo \"alias gp='find . -maxdepth 3 -name .git -type d | rev | cut -c 6- | rev | xargs -I {} git -C {} pull'\" >> ~/.bashrc"
sh -c "echo \"alias nt='nautilus .'\" >> ~/.bashrc"

sh -c "echo \"alias cw='cd ~/$name_catkin_workspace'\" >> ~/.bashrc"
sh -c "echo \"alias cs='cd ~/$name_catkin_workspace/src'\" >> ~/.bashrc"
sh -c "echo \"alias cm='cd ~/$name_catkin_workspace && catkin_make && rospack profile && source ~/$name_catkin_workspace/devel/setup.bash'\" >> ~/.bashrc"
sh -c "echo \"alias ct='cd ~/$name_catkin_workspace && catkin_make run_tests roslint && rospack profile && source ~/$name_catkin_workspace/devel/setup.bash'\" >> ~/.bashrc"
sh -c "echo \"alias ci='cd ~/$name_catkin_workspace && rosdep update && rosdep install --from-paths src --ignore-src --rosdistro $name_ros_version -r -y'\" >> ~/.bashrc"
sh -c "echo \"alias cr='cd ~/$name_catkin_workspace && rm -rf build devel'\" >> ~/.bashrc"

sh -c "echo \"alias pe='env | grep ROS'\" >> ~/.bashrc"

sh -c "echo \"alias ka='killall -9'\" >> ~/.bashrc"
sh -c "echo \"alias kg='killall -9 gazebo & killall -9 gzserver  & killall -9 gzclient'\" >> ~/.bashrc"

sh -c "echo \"source /opt/ros/$name_ros_version/setup.bash\" >> ~/.bashrc"
sh -c "echo \"source ~/$name_catkin_workspace/devel/setup.bash\" >> ~/.bashrc"

sh -c "echo \"export ROS_MASTER_URI=http://localhost:11311\" >> ~/.bashrc"
sh -c "echo \"export ROS_HOSTNAME=localhost\" >> ~/.bashrc"
sh -c "echo \"export TURTLEBOT3_MODEL=burger\" >> ~/.bashrc"
sh -c "echo \"alias tb3='ssh ubuntu@0.0.0.0' \" >> ~/.bashrc"

exec bash

echo "[Complete!!!]"
exit 0
